tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/LinkerFrontend.scala

```diff
@@ -0,0 +1,112 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js tools             **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2015, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package org.scalajs.core.tools.linker.frontend
+
+import org.scalajs.core.tools.logging.Logger
+import org.scalajs.core.tools.io.VirtualScalaJSIRFile
+
+import org.scalajs.core.tools.sem.Semantics
+import org.scalajs.core.tools.javascript.ESLevel
+
+import org.scalajs.core.tools.linker.LinkingUnit
+import org.scalajs.core.tools.linker.analyzer.SymbolRequirement
+import org.scalajs.core.tools.linker.frontend.optimizer.{GenIncOptimizer, IncOptimizer}
+
+/** The frontend of the Scala.js linker. Produces a [[LinkingUnit]]
+ *
+ *  You probably want to use an instance of [[linker.Linker]], rather than this
+ *  low-level class.
+ * 
+ *  Attention: [[LinkerFrontend]] does not cache the IR input. It is advisable to do
+ *  so, unless all IR is already in memory.
+ */
+final class LinkerFrontend(
+    val semantics: Semantics,
+    val esLevel: ESLevel,
+    val withSourceMap: Boolean,
+    config: LinkerFrontend.Config,
+    optimizerFactory: GenIncOptimizer.OptimizerFactory) {
+
+  private[this] val linker: BaseLinker =
+    new BaseLinker(semantics, esLevel, withSourceMap)
+
+  private[this] val optimizer: GenIncOptimizer =
+    if (config.disableOptimizer) null
+    else optimizerFactory(semantics, esLevel, withSourceMap)
```

Comment:
This is stupid. Why would we require someone to pass in an `optimizerFactory` and a flag? Make `optimizerFactory` an `Option` instead, drop the `disableOptimizerFlag` here. Add it to `Linker.apply`.